File: core/src/vct/col/ast/FunctionType.java
// -*- tab-width:2 ; indent-tabs-mode:nil -*-
package vct.col.ast;

import java.util.*;

public class FunctionType extends Type {

  @Override
  public <R,A> R accept_simple(ASTMapping1<R,A> map,A arg){
    return map.map(this,arg);
  }

  private Type args[];
  private Type result;
  public FunctionType(Type args[],Type result){
    this.args=Arrays.copyOf(args,args.length);
    this.result=result;
  }
  public FunctionType(ArrayList<Type> args,Type result){
    this.args=args.toArray(new Type[0]);
    this.result=result;
  }

  public int getArity(){ return args.length; }
  
  public Type getResult(){ return result; }
  
  public Type getArgument(int i){ return args[i]; }
  
  @Override
  public ASTNode zero(){
    return null;
  }
  
  @Override
  public <T> void accept_simple(ASTVisitor<T> visitor){
    try {
      visitor.visit(this);
    } catch (Throwable t){
      if (thrown.get()!=t){
        System.err.printf("Triggered by %s:%n",getOrigin());
        thrown.set(t);
    }
      throw t;
    }
  }
  
  @Override
  public <T> T accept_simple(ASTMapping<T> map){
    try {
      return map.map(this);
    } catch (Throwable t){
      if (thrown.get()!=t){
        System.err.printf("Triggered by %s:%n",getOrigin());
        thrown.set(t);
    }
      throw t;
    }
  }
 
  @Override
  public <T> T accept_simple(TypeMapping<T> map){
    try {
      return map.map(this);
    } catch (Throwable t){
      if (thrown.get()!=t){
        System.err.printf("Triggered by %s:%n",getOrigin());
        thrown.set(t);
     }
      throw t;
    }
  }

  @Override
  public boolean supertypeof(ProgramUnit context, Type t) {
    // TODO Auto-generated method stub
    return false;
  }

  public boolean equals(Object o){
    if (o instanceof FunctionType){
      FunctionType t=(FunctionType)o;
      if (t.args.length!=args.length) return false;
      for(int i=0;i<args.length;i++){
        if(!args[i].equals(t.args[i])) return false;
      }
      return result.equals(t.result);
    } else {
      return false;
    }
  }
}

File: vercors/src/main/java/vct/col/ast/type/FunctionType.scala
package vct.col.ast.`type`

import scala.collection.JavaConverters._
import vct.col.ast.stmt.decl.ProgramUnit
import vct.col.ast.util.{ASTVisitor, TypeMapping}
import vct.col.util.{ASTMapping, ASTMapping1, VisitorHelper}

/**
 * AST node representing function types (`T_0*...*T_n -> T`), where `params` is the
 * list of parameter types (`T_0,...,T_n`) and `result` the return type (`T`).
 *
 * @author sccblom, whmoortwijn
 */
case class FunctionType(val params:List[Type], val result:Type) extends Type with VisitorHelper {
  require(params != null, "The parameter list is null")
  require(result != null, "Function types should have a result type")

  /** Constructs a new function type from an array of parameter types */
  def this(params:Array[Type], result:Type) = this(params.toList, result)

  /** Constructs a new function type from Java constructs (for Java interoperability) */
  def this(params:java.util.List[Type], result:Type) = this(params.asScala.toList, result)

  /** Constructs a new unary function type */
  def this(param:Type, result:Type) = this(List(param), result)

  /** Provides a Java wrapper (as `java.util.List`) for the list of parameter types. */
  def paramsJava = params.asJava

  override def equals(other:Any) = other match {
    case FunctionType(`params`, `result`) => true
    case _ => false
  }

  override def accept_simple[T,A](m:ASTMapping1[T,A], arg:A) = m.map(this, arg)
  override def accept_simple[T](v:ASTVisitor[T]) = handle_standard(() => v.visit(this))
  override def accept_simple[T](m:ASTMapping[T]) = handle_standard(() => m.map(this))
  override def accept_simple[T](m:TypeMapping[T]) = handle_standard(() => m.map(this))
  override def supertypeof(context:ProgramUnit, otherType:Type) = false
  override val zero = null

  override def debugTreeChildrenFields: Iterable[String] = Seq("params", "result", "args")
  override def debugTreePropertyFields: Iterable[String] = Seq()
}
